File: GP Binary Classification/Python/GPCbinary.py
import numpy as np
import matplotlib.pyplot as plt
import math

# FUNCTIONS
def ARDkernel(x1,x2,para):
    sf = para[0]
    dim = len(x1)
    s = 0
    for i in range(dim):
        s = s + para[i+1]*(x1[i]-x2[i])**2
    k = sf**2 * math.exp(-0.5*s)
    return k

def PDe_ARDkernel_sf(x1,x2,para):
    sf = para[0]
    dim = len(x1)
    s = 0
    for i in range(dim):
        s = s + para[i + 1] * (x1[i] - x2[i]) ** 2
    k = 2 * sf * math.exp(-0.5 * s)
    return k

def PDe_ARDkernel_alpha(x1,x2,para,q):
    sf = para[0]
    dim = len(x1)
    s = 0
    for i in range(dim):
        s = s + para[i + 1] * (x1[i] - x2[i]) ** 2
    k = -0.5 * sf ** 2 * math.exp(-0.5 * s) * (x1[q-1] - x2[q-1])**2
    return k

def K_CovMatrix(N,trainx,para):
    K = np.zeros((N, N))
    for i in range(N):
        for j in range(N):
            K[i][j] = ARDkernel(trainx[:, [i]],trainx[:, [j]], para)
    return K

def Kstar_CovMatrix(N,M,trainx,testx,para):
    K = np.zeros((N, M))
    for i in range(N):
        for j in range(M):
            K[i][j] = ARDkernel(trainx[:, [i]],testx[:, [j]], para)
    return K

def Kstar2_CovMatrix(M,testx,para):
    K = np.zeros((M, M))
    for i in range(M):
        for j in range(M):
            K[i][j] = ARDkernel(testx[:, [i]],testx[:, [j]], para)
    return K

def PDeK_CovMatrix_sf(N,trainx,para):
    K = np.zeros((N, N))
    for i in range(N):
        for j in range(N):
            K[i][j] = PDe_ARDkernel_sf(trainx[:, [i]],trainx[:, [j]], para)
    return K

def PDeK_CovMatrix_alpha(N,trainx,para,q):
    K = np.zeros((N, N))
    for i in range(N):
        for j in range(N):
            K[i][j] = PDe_ARDkernel_alpha(trainx[:, [i]],trainx[:, [j]], para,q)
    return K


def Logistic(y,f):
    return 1/(1+math.exp(-y*f))

def LogLogistic(y,f):
    return -math.log(1+math.exp(-y*f))

def De1_LogLogistic(y,f):
    return (y+1)/2 - 1/(1+math.exp(-f))

def De2_LogLogistic(f):
    return -1/(1+math.exp(-f))*(1-1/(1+math.exp(-f)))

def De3_LogLogistic(f):
    return 2*math.exp(-f)/(1+math.exp(-f))**3 - math.exp(-f)/(1+math.exp(-f))**2

def Neg_Log_Posterior(F,args):
    (trainy, trainx, Para) = args
    N = len(trainy)
    K = K_CovMatrix(N, trainx, Para)
    loglike = 0
    for i in range(N):
        loglike = loglike + LogLogistic(trainy[i], F[i])

    return 0.5*np.transpose(F)@np.linalg.inv(K)@F + 0.5*math.log(np.linalg.det(K)) + 0.5*N*math.log(2*np.pi) - loglike

def PDe1_Neg_Log_Posterior(F,args):
    (trainy, trainx, Para) = args
    N = len(trainy)
    K = K_CovMatrix(N, trainx, Para)
    de1_log_like_list = []
    for i in range(N):
        de1_log_like_list.append(De1_LogLogistic(trainy[i], F[i]))
    de1_log_like_list = np.array(de1_log_like_list)
    return np.linalg.inv(K)@F - np.transpose(de1_log_like_list)

def PDe2_Neg_Log_Posterior(F,args):
    (trainy, trainx, Para) = args
    N = len(trainy)
    K = K_CovMatrix(N, trainx, Para)
    de2_log_like_list = []
    for i in range(N):
        de2_log_like_list.append(De2_LogLogistic(F[i]))
    W = - np.diag(de2_log_like_list)
    return np.linalg.inv(K) + W


def Neg_LogMar_Likelihood(Para,args):
    (trainy, trainx, F) = args

    N = len(trainy)
    K = K_CovMatrix(N, trainx, Para)
    de2_log_like_list = []
    for i in range(N):
        de2_log_like_list.append(De2_LogLogistic(F[i]))
    W = - np.diag(de2_log_like_list)
    B = np.identity(N) + np.sqrt(W) @ K @ np.sqrt(W)
    loglike = 0
    for i in range(N):
        loglike = loglike + LogLogistic(trainy[i], F[i])

    output = 0.5*np.transpose(F)@np.linalg.inv(K)@F - loglike + 0.5*math.log(np.linalg.det(B))
    '''
    de2_log_like_list = []
    N = len(trainy)
    for i in range(N):
        de2_log_like_list.append(De2_LogLogistic(F[i]))
    W = - np.diag(de2_log_like_list)
    K = K_CovMatrix(N, trainx, Para)
    B = np.identity(N) + np.sqrt(W) @ K @ np.sqrt(W)
    L = np.linalg.cholesky(B)
    de1_log_like_list = []
    for i in range(N):
        de1_log_like_list.append(De1_LogLogistic(trainy[i], F[i]))
    de1_log_like_list = np.array(de1_log_like_list)
    b = W @ F + np.transpose(de1_log_like_list)
    a = b - np.linalg.solve(np.sqrt(W) @ np.transpose(L), (np.linalg.solve(L, np.sqrt(W) @ K @ b)))
    loglike = 0
    for i in range(N):
        loglike = loglike + LogLogistic(trainy[i],F[i])

    return 0.5* np.transpose(a)@F - loglike + sum(np.log(np.diag(L)))
    '''
    return output

def PDe_Neg_LogMar_Likelihood(Para,args):
    (trainy, trainx, F) = args
    N = len(trainy)
    NumofPara = len(Para)
    de2_log_like_list = []
    for i in range(N):
        de2_log_like_list.append(De2_LogLogistic(F[i]))
    W = - np.diag(de2_log_like_list)
    K = K_CovMatrix(N, trainx, Para)
    B = np.identity(N) + np.sqrt(W) @ K @ np.sqrt(W)
    DrK_sf = PDeK_CovMatrix_sf(N, trainx, Para)
    output = []
    output.append( -0.5*np.transpose(F)@(np.linalg.inv(K)@DrK_sf@np.linalg.inv(K))@F + 0.5*np.trace(np.linalg.inv(B)@W@DrK_sf) )
    for i in range(1,NumofPara):
        Drk_alpha = PDeK_CovMatrix_alpha(N,trainx,Para,i)
        output.append( -0.5*np.transpose(F)@(np.linalg.inv(K)@Drk_alpha@np.linalg.inv(K))@F + 0.5*np.trace(np.linalg.inv(B)@W@Drk_alpha) )


    '''
    de2_log_like_list = []
    de3_log_like_list = []
    N = len(trainy)
    for i in range(N):
        de2_log_like_list.append(De2_LogLogistic(F[i]))
        de3_log_like_list.append(De3_LogLogistic(F[i]))
    de3_log_like_list = np.array(de3_log_like_list)
    W = - np.diag(de2_log_like_list)
    K = K_CovMatrix(N, trainx, Para)
    B = np.identity(N) + np.sqrt(W) @ K @ np.sqrt(W)
    L = np.linalg.cholesky(B)
    de1_log_like_list = []
    for i in range(N):
        de1_log_like_list.append(De1_LogLogistic(trainy[i], F[i]))
    de1_log_like_list = np.array(de1_log_like_list)
    b = W @ F + np.transpose(de1_log_like_list)
    a = b - np.linalg.solve(np.sqrt(W) @ np.transpose(L), (np.linalg.solve(L, np.sqrt(W) @ K @ b)))

    R = np.linalg.solve(np.sqrt(W)@np.transpose(L), np.linalg.solve(L,np.sqrt(W)) )
    C = np.linalg.solve(L,np.sqrt(W)@K)
    s2 = -0.5*np.diag(np.diag(K)-np.diag(np.transpose(C)@C))@np.transpose(de3_log_like_list)

    NumofPara = len(Para)
    output = []
    for i in range(NumofPara):
        if i == 0:
            C = PDeK_CovMatrix_sf(N,trainx,Para)
        else:
            C = PDeK_CovMatrix_alpha(N,trainx,Para,i)
        s1 = 0.5*np.transpose(a)@C@a - 0.5*np.trace(R*C)
        bb = C @ np.transpose(de1_log_like_list)
        s3 = bb - K @ R @ bb
        output.append(-s1 - np.transpose(s2)@s3)
    '''
    return np.array(output)


def OPT_GradientDescent(Fun,J,x0,MaxIter,Termination,Step,args):
    fvals = []
    iter = 1
    x = x0
    fvals.append(Fun(x,args))
    flag = 1
    while (iter <= MaxIter) and (flag == 1):
        iter = iter + 1
        delta = Step*J(x,args)
        x = x - delta
        fvals.append(Fun(x, args))
        #print("The negative log marginal likelihood is {}.".format(fvals[-1]))
        if np.linalg.norm(delta)<Termination:
            flag = 0

    return x,fvals[-1]

def OPT_Newton(Fun,J,H,x0,MaxIter,Termination,Step,args):
    fvals = []
    iter = 1
    x = x0
    dim = len(x)
    fvals.append(Fun(x,args))
    flag = 1
    while (iter <= MaxIter) and (flag == 1):
        iter = iter + 1
        e,v = np.linalg.eig(H(x,args))
        if any(e<0):
            MU = -min(e)+0.001
        else:
            MU = 0
        delta = Step* np.linalg.inv(H(x,args)+MU*np.identity(dim)) @ J(x,args)
        x = x - delta
        fvals.append(Fun(x, args))
        #print("The negative log posterior is {}.".format(fvals[-1]))
        if np.linalg.norm(delta)<Termination:
            flag = 0

    return x, fvals[-1]

# generate data

NumofClass = 2
NumofTrain_each = 25
NumofTrain = NumofTrain_each * NumofClass
NumofTest_each = 5
NumofTest = NumofTest_each * NumofClass

Mu_c1 = [2,5]    # mean of classes
Mu_c2 = [5,-1]

Cov_c1 = [[6,0],[0,10]]   # covariance of classes
Cov_c2 = [[10,5],[5,10]]

x_c1, y_c1 = np.random.multivariate_normal(Mu_c1, Cov_c1, NumofTrain_each).T
x_c2, y_c2 = np.random.multivariate_normal(Mu_c2, Cov_c2, NumofTrain_each).T

TrainX = np.array([x_c1.tolist()+x_c2.tolist(),y_c1.tolist()+y_c2.tolist()])
TrainY = [1]*NumofTrain_each + [-1]*NumofTrain_each

x_c1_te, y_c1_te = np.random.multivariate_normal(Mu_c1, Cov_c1, NumofTest_each).T
x_c2_te, y_c2_te = np.random.multivariate_normal(Mu_c2, Cov_c2, NumofTest_each).T

TestX = np.array([x_c1_te.tolist()+x_c2_te.tolist(),y_c1_te.tolist()+y_c2_te.tolist()])
TestY = [1]*NumofTest_each + [-1]*NumofTest_each

fig1 = plt.figure(figsize=(6,6))
plt.plot(x_c1, y_c1, 'oc')
plt.plot(x_c2, y_c2, 'oy')
plt.plot(TestX[0], TestX[1], '*r',Markersize=10)



#############################################GP Binary Classification
Dim = TrainX.shape[0]   # dimension of inputs
NumofIteration = 5
InitialPara = [1] + [1]*Dim  # initial hyperparameters
InitialF = [0]*NumofTrain    # initial latent function

Para = InitialPara
F = InitialF

while NumofIteration>0:
    # part1: optimize posterior
    F, fun = OPT_Newton(Neg_Log_Posterior,PDe1_Neg_Log_Posterior,PDe2_Neg_Log_Posterior,F,1000,1e-6,0.2,(TrainY,TrainX,Para))
    print("The negative log posterior is {}.".format(fun))
    print("----------")
    '''
    flag = 1
    # part1: optimize posterior 
    while flag==1:
        de2_log_like_list = []
        for i in range(NumofTrain):
            de2_log_like_list.append(De2_LogLogistic(F[i]))
        W = - np.diag(de2_log_like_list)
        K = K_CovMatrix(NumofTrain,TrainX,Para)
        B = np.identity(NumofTrain) + np.sqrt(W)@K@np.sqrt(W)
        L = np.linalg.cholesky(B)
        de1_log_like_list = []
        for i in range(NumofTrain):
            de1_log_like_list.append(De1_LogLogistic(TrainY[i],F[i]))
        de1_log_like_list = np.array(de1_log_like_list)
        b = W@F + np.transpose(de1_log_like_list)
        a = b - np.linalg.solve( np.sqrt(W)@np.transpose(L), (np.linalg.solve(L,np.sqrt(W)@K@b)) )

        Fnew = K@a

        delta = np.linalg.norm(Fnew-F)
        if delta<1e-2:
            flag = 0
        F = Fnew
    '''



    # part2: optimize marginal likelihood
    '''
    opts = {'maxiter': 1,
            'disp': True,  # non-default value.
            'gtol': 1e-1,
            'norm': np.inf,  # default value.
            'eps': 1.4901161193847656e-08}  # default value.
    Para,fun = optimize.minimize(Neg_LogMar_Likelihood, Para, jac=PDe_Neg_LogMar_Likelihood, args=(TrainY,TrainX,F), method="Newton-CG", options=opts)
    '''
    #Para, fun = sp.optimize.fmin_cg(Neg_LogMar_Likelihood, Para, fprime=PDe_Neg_LogMar_Likelihood, args=(TrainY,TrainX,F),gtol=1e-1,maxiter=1)
    #Para, fun = sp.optimize.fmin_bfgs(Neg_LogMar_Likelihood, Para, fprime=PDe_Neg_LogMar_Likelihood,args=(TrainY, TrainX, F), gtol=1e-1, maxiter=1)
    Para, fun = OPT_GradientDescent(Neg_LogMar_Likelihood, PDe_Neg_LogMar_Likelihood, Para, 1000, 1e-6, 0.005, (TrainY,TrainX,F))
    #print("The negative log marginal likelihood is {}.".format(fun))
    print("The negative log marginal likelihood is {}.".format(fun))
    print("----------")

    NumofIteration = NumofIteration - 1


# prediction
de1_log_like_list = []
for i in range(NumofTrain):
    de1_log_like_list.append(De1_LogLogistic(TrainY[i],F[i]))
de1_log_like_list = np.array(de1_log_like_list)
de2_log_like_list = []
for i in range(NumofTrain):
    de2_log_like_list.append(De2_LogLogistic(F[i]))
W = - np.diag(de2_log_like_list)
K = K_CovMatrix(NumofTrain,TrainX,Para)
Kstar = Kstar_CovMatrix(NumofTrain,NumofTest,TrainX,TestX,Para)
Kstar2 = Kstar2_CovMatrix(NumofTest,TestX,Para)
B = np.identity(NumofTrain) + np.sqrt(W) @ K @ np.sqrt(W)
L = np.linalg.cholesky(B)
Fstar_mean = np.transpose(Kstar) @ np.transpose(de1_log_like_list)
v = np.linalg.solve(L, np.sqrt(W)@Kstar)
Fstar_cov = Kstar2 - np.transpose(v) @ v
PI = 1/(1 + np.exp(-Fstar_mean))

print(PI)

for i in range(NumofTest):
    plt.text(TestX[0][i], TestX[1][i], PI[i], fontsize=12)

plt.show()

File: GP Binary Classification/MATLAB/GPC_binary.m
function [output] = GPC_binary(X,Y,COVF,InitialPara,loglik,Dr1_loglik,Dr2_loglik,Xstar)
% Training data X(d*N)
%

% Number and dimension of samples
NumTrain = length(Y);
NumTest = length(Xstar(1,:));
Dim = length(X(:,1));

% Number of iteration
I = 10;

% initial hyperparameters
theta = InitialPara;

% initial latent function
f = zeros(NumTrain,1);

% % covariance matrix of training data
% for i = 1:NumTrain
%     for j = 1:NumTrain
%         K(i,j) = COVF(X(:,i),X(:,j),sf,l);
%     end
% end

temp = 100;
for count = 1:I
    
% find f_hat by minimizing the negative posterior -p(f|X,y)
Neg_Log_Post = @(f) 0.5*f'*inv(K(theta,X,COVF))*f + 0.5*log(det(K(theta,X,COVF))) + 0.5*NumTrain*log(2*pi) - Lik(f,loglik,Y);
Dr1_Neg_Log_Post = @(f) inv(K(theta,X,COVF))*f - Dr1Lik(f,Dr1_loglik,Y);
Dr2_Neg_Log_Post = @(f) inv(K(theta,X,COVF)) - Dr2Lik(f,Dr2_loglik,Y);

[f_hat, neg_posterior_f] = Fun_mini_Newton(Neg_Log_Post,Dr1_Neg_Log_Post,Dr2_Neg_Log_Post,f,1e3,1e-6,0.1);

% optimize hyperparameters by minimizing the negative marginal likelihood -q(y|X,theta)
W = -Dr2Lik(f_hat,Dr2_loglik,Y);
B = @(theta) eye(NumTrain) + sqrt(W)*K(theta,X,COVF)*sqrt(W);
Neg_Log_Marg = @(theta) 0.5*f_hat'*inv(K(theta,X,COVF))*f_hat - Lik(f_hat,loglik,Y) + 0.5*log(det(eye(NumTrain)+sqrt(W)*K(theta,X,COVF)*sqrt(W)));
Dr1_Neg_Log_Marg = @(theta) Fun_Dr1_Neg_Log_Marg(theta,X,f_hat,W,B);
Dr2_Neg_Log_Marg = @(theta) Fun_Dr2_Neg_Log_Marg(theta,X,f_hat,W,B);

[theta_hat, neg_marginal] = Fun_mini_gredd(Neg_Log_Marg,Dr1_Neg_Log_Marg,theta,1e3,1e-6,0.001)

%[theta_hat, marginal] = Fun_mini_Newton(Neg_Log_Marg,Dr1_Neg_Log_Marg,Dr2_Neg_Log_Marg,InitialPara,1e3,1e-6,0.001);

if temp-neg_marginal<0.1
    break;
end

f = f_hat;
theta = theta_hat;

temp = neg_marginal;
end

% Predict the mean and covariance of test latent variables
W = -Dr2Lik(f,Dr2_loglik,Y);
Bmatrix = B(theta);
L = chol(Bmatrix);
fstar_mean = Kstar(theta,X,Xstar,COVF)'*Dr1Lik(f,Dr1_loglik,Y);
v = L\(sqrt(W)*Kstar(theta,X,Xstar,COVF));
fstar_cov = Kstar2(theta,Xstar,COVF) - v'*v;

% Compute predictive probabilites
%MAP prediction
output = 1./(ones(NumTest,1)+exp(-fstar_mean));

end

function output = Lik(f,loglik,Y)
output = 0;
for fi = 1:length(f)
    output = output + loglik(Y(fi),f(fi));
end
end

function output = Dr1Lik(f,Dr1_loglik,Y)
for fi = 1:length(f)
    output(fi,1) = Dr1_loglik(Y(fi),f(fi));
end
end

function output = Dr2Lik(f,Dr2_loglik,Y)
for fi = 1:length(f)
    output(fi,1) = Dr2_loglik(Y(fi),f(fi));
end
output = diag(output);
end

function output = K(theta,X,COVF)
for fi = 1:length(X(1,:))
    for fj = 1:length(X(1,:))
        output(fi,fj) = COVF(X(:,fi),X(:,fj),theta(1),exp(theta(2)));
    end
end
end

function output = Kstar(theta,X,Xstar,COVF)
for fi = 1:length(X(1,:))
    for fj = 1:length(Xstar(1,:))
        output(fi,fj) = COVF(X(:,fi),Xstar(:,fj),theta(1),exp(theta(2)));
    end
end
end

function output = Kstar2(theta,Xstar,COVF)
for fi = 1:length(Xstar(1,:))
    for fj = 1:length(Xstar(1,:))
        output(fi,fj) = COVF(Xstar(:,fi),Xstar(:,fj),theta(1),exp(theta(2)));
    end
end
end

function output = Fun_Dr1_Neg_Log_Marg(theta,X,f_hat,W,B)
sf = theta(1);
l = exp(theta(2));
for fi = 1:length(X(1,:))
    for fj = 1:length(X(1,:))
        K(fi,fj) = sf^2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr1K_sf(fi,fj) = 2*sf*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr1K_logl(fi,fj) = sf^2*norm(X(:,fi)-X(:,fj))^2/l^2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
    end
end
output(1,1) = -0.5*f_hat'*(inv(K)*Dr1K_sf*inv(K))*f_hat + 0.5*trace(inv(B(theta))*W*Dr1K_sf);
output(2,1) = -0.5*f_hat'*(inv(K)*Dr1K_logl*inv(K))*f_hat + 0.5*trace(inv(B(theta))*W*Dr1K_logl);
end

function output = Fun_Dr2_Neg_Log_Marg(theta,X,f_hat,W,B)
sf = theta(1);
l = exp(theta(2));
for fi = 1:length(X(1,:))
    for fj = 1:length(X(1,:))
        K(fi,fj) = sf^2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr1K_sf(fi,fj) = 2*sf*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr1K_logl(fi,fj) = sf^2*norm(X(:,fi)-X(:,fj))^2/l^2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr2K_sf(fi,fj) = 2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr2K_logl(fi,fj) = sf^2*norm(X(:,fi)-X(:,fj))^2*(-2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2)/l^2+norm(X(:,fi)-X(:,fj))^2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2)/l^4);
        Dr2K_sf2l(fi,fj) = 2*sf*norm(X(:,fi)-X(:,fj))^2/l^2*exp(-0.5/l^2*norm(X(:,fi)-X(:,fj))^2);
        Dr2K_l2sf(fi,fj) = Dr2K_sf2l(fi,fj);
    end
end
output(1,1) = 0.5*trace(-W*inv(B(theta))*W*Dr1K_sf*inv(B(theta))*Dr1K_sf+W*inv(B(theta))*Dr2K_sf) -...
    -0.5*f_hat'*inv(K)*(-2*Dr1K_sf*inv(K)*Dr1K_sf+Dr2K_sf)*inv(K)*f_hat;
output(1,2) = 0.5*trace(-W*inv(B(theta))*W*Dr1K_logl*inv(B(theta))*Dr1K_sf+W*inv(B(theta))*Dr2K_sf2l) -...
    -0.5*f_hat'*inv(K)*(-Dr1K_logl*inv(K)*Dr1K_sf+Dr2K_sf2l-Dr1K_sf*inv(K)*Dr1K_logl)*inv(K)*f_hat;
output(2,1) = 0.5*trace(-W*inv(B(theta))*W*Dr1K_sf*inv(B(theta))*Dr1K_logl+W*inv(B(theta))*Dr2K_l2sf) -...
    -0.5*f_hat'*inv(K)*(-Dr1K_sf*inv(K)*Dr1K_logl+Dr2K_l2sf-Dr1K_logl*inv(K)*Dr1K_sf)*inv(K)*f_hat;
output(2,2) = 0.5*trace(-W*inv(B(theta))*W*Dr1K_logl*inv(B(theta))*Dr1K_logl+W*inv(B(theta))*Dr2K_logl) -...
    -0.5*f_hat'*inv(K)*(-2*Dr1K_logl*inv(K)*Dr1K_logl+Dr2K_logl)*inv(K)*f_hat;

end

